### PR TITLE
Add support for trailing commas, fixes #657

### LINKED ISSUE #657: Trailing commas in 2.12.2
PR was just merged https://github.com/scala/scala/pull/5245.

We need to allow trailing commas in 2.12.2 but not previous version. much like literal type syntax

### PR DESCRIPTION
**Pull request by olafurpg**
cc/ https://github.com/scala/scala-dev/issues/299

### CODE REVIEW

File: scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala [hunk 1]
```diff
@@ -105,6 +109,7 @@ package object dialects {
     def allowOrTypes = Scala210.allowOrTypes
     def allowTraitParameters = Scala210.allowTraitParameters
     def allowLiteralTypes: Boolean = Scala210.allowLiteralTypes
+    def allowTrailingComma: Boolean = Scala210.allowLiteralTypes
```
Comment (dwijnand) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: is the RHS here intentional?

Comment (olafurpg) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: Yes. This setup is quite verbose but it reads easily.

File: scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala [hunk 2]
```diff
@@ -173,6 +181,7 @@ package object dialects {
     def allowOrTypes = Scala211.allowOrTypes
     def allowTraitParameters = Scala211.allowTraitParameters
     def allowLiteralTypes: Boolean = false // Scheduled to be included in 2.12.2
+    def allowTrailingComma: Boolean = true // Schedule to be included in 2.12.2
```
Comment (dwijnand) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: both allowTrailingComma and allowLiteralTypes and scheduled for 2.12.2, but one is false and the other is true?

File: scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala [hunk 3]
```diff
@@ -122,6 +127,7 @@ package object dialects {
     def allowOrTypes = Scala210.allowOrTypes
     def allowTraitParameters = Scala210.allowTraitParameters
     def allowLiteralTypes: Boolean = Scala210.allowLiteralTypes
+    def allowTrailingComma: Boolean = Scala210.allowLiteralTypes
```
Comment (dwijnand) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: So this dialect is what we'll use in sbt 1, which will use Scala 2.12.2+ and therefore _will_ allow trailing commas. But in sbt between 0.13.7 and 1.0.0 (`[0.13.7,1.0.0)`) it's not. Can we model _that_ here?

File: scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala (hunk 1 above)
Comment (dwijnand) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: You sure you don't mean `Scala210. allowTrailingComma`?

File: scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala (hunk 2 above)
Comment (olafurpg) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: Good question. I think both should be true, but don't want to change allowLiteralTypes in this PR.

File: scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala (hunk 3 above)
Comment (olafurpg) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: There will be a custom Sbt100 (or something like that, maybe `Sbt1`) dialect where we can set all the fields as desired.

Comment (dwijnand) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: ok cool

File: scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala (hunk 1 above)
Comment (olafurpg) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: Oh boy -_- good catch, thanks! Fixed now

Comment (dwijnand) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: Copy-pasta, the arch-enemy of "quite verbose but it reads easily" :-P

Comment (olafurpg) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: Haha you're totally right. I actually hate this, was trying to find something nice to say about it. Opened https://github.com/scalameta/scalameta/issues/661

File: scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala (hunk 2 above)
Comment (xeno-by) [inline scalameta/dialects/src/main/scala/scala/meta/dialects/package.scala]: Let's make both false and then update both to true when 2.12 is released (see https://github.com/scalameta/scalameta/issues/645).

**Comment by olafurpg** [comment]
How am I supposed to test the functionality if no dialect supports trailling commas? By having an "AllowEverything" dialect ;) (I've wanted this forever!) I took the liberty to add such a dialect.

**Comment by xeno-by** [comment]
@olafurpg Lol, I see :)

**Pull request merged**